lib/src/clojure_lsp/feature/diagnostics.clj

```diff
@@ -185,47 +184,64 @@
+                    (into #{}
+                          (comp
+                            q/xf-all-keyword-usages
+                            (map q/kw-signature))
+                          (:analysis project-db))
+                    #{})
         kw-used? (fn [kw-def]
-                   (contains? kw-usages (q/kw-signature kw-def)))
-        findings (->> (concat (remove var-used? var-definitions)
-                              (remove kw-used? kw-definitions))
-                      (map (fn [unused-var]
-                             (unused-public-var->finding unused-var config))))]
-    (doseq [finding findings] ;; side-effect to register findings
-      (reg-finding! finding))
-    (group-by :filename findings)))
+                   (contains? kw-usages (q/kw-signature kw-def)))]
+    (->> (concat (remove var-used? var-definitions)
+                 (remove kw-used? kw-definitions))
+         (map (fn [unused-var]
+                (unused-public-var->finding unused-var kondo-config))))))
 
 (defn ^:private file-var-definitions [project-db filename]
   (q/find-var-definitions project-db filename false))
 (def ^:private file-kw-definitions q/find-keyword-definitions)
 (def ^:private all-var-definitions q/find-all-var-definitions)
 (def ^:private all-kw-definitions q/find-all-keyword-definitions)
 
-(defn custom-lint-project!
-  [db kondo-ctx]
+(defn project-findings
+  [db kondo-config]
   (let [project-db (q/db-with-internal-analysis db)]
-    (lint-defs! (all-var-definitions project-db)
-                (all-kw-definitions project-db)
-                project-db kondo-ctx)))
+    (unused-public-vars (all-var-definitions project-db)
+                        (all-kw-definitions project-db)
+                        project-db kondo-config)))
 
-(defn custom-lint-files!
-  [filenames db kondo-ctx]
+(defn files-findings
+  [filenames db kondo-config]
   (let [project-db (q/db-with-internal-analysis db)
         files-db (update project-db :analysis select-keys filenames)]
-    (lint-defs! (all-var-definitions files-db)
-                (all-kw-definitions files-db)
-                project-db kondo-ctx)))
+    (unused-public-vars (all-var-definitions files-db)
+                        (all-kw-definitions files-db)
+                        project-db kondo-config)))
 
-(defn custom-lint-file!
-  [filename db kondo-ctx]
+(defn file-findings
+  [filename db kondo-config]
   (let [project-db (q/db-with-internal-analysis db)]
-    (lint-defs! (file-var-definitions project-db filename)
-                (file-kw-definitions project-db filename)
-                project-db kondo-ctx)))
+    (unused-public-vars (file-var-definitions project-db filename)
+                        (file-kw-definitions project-db filename)
+                        project-db kondo-config)))
+
+(defn ^:private finalize-findings! [findings reg-finding!]
+  (run! reg-finding! findings)
```

Comment:
Looks better, if we stop using custom-lint-fn in the future, it's easier to remove this code here